File: leetcode/middle/2594/2594.cpp
#define ll long long
class Solution {
public:
    inline bool chk(ll x,vector<int>& ranks,int c)
    {
        ll n = ranks.size();
        for(int i = 0;i<n;i++)
        {
            ll lim = sqrt(x/ranks[i]);
            c-=lim;
            if (c<=0) return 1;
        }
        return 0;
    }
    long long repairCars(vector<int>& ranks, int cars) {
        ll l = 0,r = 1e18+7;
        ll ans = 0;
        while(l<=r)
        {
            ll mid = (l+r)/2;
            if(chk(mid,ranks,cars)) {ans = mid;r = mid-1;}
            else l = mid+1;
        }
        return ans;
    }
};

File: leetcode/middle/2594/2594.py
from typing import List
from math import sqrt
class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        n : int = len(ranks)
        l : int = 0 
        r : int = 1e18+7
        ans: int  = 0
        def check(mid, ranks, cars) -> bool:
            cnt : int = 0
            for i in range(n):
                cnt += int (sqrt(mid//ranks[i]))
                if cnt >= cars:
                    return True
            return False
        while l <= r:
            mid = (l+r)//2
            if check(mid, ranks, cars):
                ans = mid
                r = mid-1
            else:
                l = mid+1
        return ans